openapi-codegen/core/src/main/scala/sttp/tapir/codegen/ClassDefinitionGenerator.scala

```diff
@@ -3,96 +3,190 @@ package sttp.tapir.codegen
-      doc.components.toSeq
-        .flatMap(_.schemas.collect { case (name, _: OpenapiSchemaEnum) => name })
+      allSchemas
+        .collect { case (name, _: OpenapiSchemaEnum) => name }
         .exists(queryParamRefs.contains)
-    val enumQuerySerdeHelper =
-      if (!generatesQueryParamEnums) ""
-      else if (targetScala3)
-        """
-          |def enumMap[E: enumextensions.EnumMirror]: Map[String, E] =
-          |  Map.from(
-          |    for e <- enumextensions.EnumMirror[E].values yield e.name.toUpperCase -> e
-          |  )
-          |
-          |def makeQueryCodecForEnum[T: enumextensions.EnumMirror]: sttp.tapir.Codec[List[String], T, sttp.tapir.CodecFormat.TextPlain] =
-          |  sttp.tapir.Codec
-          |    .listHead[String, String, sttp.tapir.CodecFormat.TextPlain]
-          |    .mapDecode(s =>
-          |      // Case-insensitive mapping
-          |      scala.util
-          |        .Try(enumMap[T](using enumextensions.EnumMirror[T])(s.toUpperCase))
-          |        .fold(
-          |          _ =>
-          |            sttp.tapir.DecodeResult.Error(
-          |              s,
-          |              new NoSuchElementException(
-          |                s"Could not find value $s for enum ${enumextensions.EnumMirror[T].mirroredName}, available values: ${enumextensions.EnumMirror[T].values.mkString(", ")}"
-          |              )
-          |            ),
-          |          sttp.tapir.DecodeResult.Value(_)
-          |        )
-          |    )(_.name)
-          |""".stripMargin
-      else
-        """def makeQueryCodecForEnum[T <: enumeratum.EnumEntry](enumName: String, T: enumeratum.Enum[T]): sttp.tapir.Codec[List[String], T, sttp.tapir.CodecFormat.TextPlain] =
-          |  sttp.tapir.Codec.listHead[String, String, sttp.tapir.CodecFormat.TextPlain]
-          |    .mapDecode(s =>
-          |      // Case-insensitive mapping
-          |      scala.util.Try(T.upperCaseNameValuesToMap(s.toUpperCase))
-          |        .fold(
-          |          _ =>
-          |            sttp.tapir.DecodeResult.Error(
-          |              s,
-          |              new NoSuchElementException(
-          |                s"Could not find value $s for enum ${enumName}, available values: ${T.values.mkString(", ")}"
-          |              )
-          |            ),
-          |          sttp.tapir.DecodeResult.Value(_)
-          |        )
-          |    )(_.entryName)
-          |""".stripMargin
+
+    def fetchJsonParamRefs(initialSet: Set[String], toCheck: Seq[OpenapiSchemaType]): Set[String] = toCheck match {
+      case Nil          => initialSet
+      case head +: tail => recursiveFindAllReferencedSchemaTypes(allSchemas)(head, initialSet, tail)
+    }
+
+    val allTransitiveJsonParamRefs = fetchJsonParamRefs(
+      jsonParamRefs,
+      jsonParamRefs.toSeq.flatMap(ref => allSchemas.get(ref.stripPrefix("#/components/schemas/")))
+    )
+
+    val maybeJsonSerdeHelpers =
```

Comment:
**question:** Shouldn't jsoniter-scala derive these automatically?

Reply:
You would expect that, wouldn't you? I haven't really used jsoniter-scala much before so perhaps there's a less ham-fisted way to do this, but without these I get a lot of `could not find implicit value for evidence parameter of type com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec[List[sttp.tapir.generated.TapirGeneratedEndpoints.MyCaseClass]]` and whatnot. My understanding is that it's expected that all top-level json formats will require an explicit serde declaration.

Reply:
There is an ability to auto-derive codecs using an in-lined compile-time configuration, like [here](https://github.com/plokhotnyuk/jsoniter-scala/blob/8b730632c8ad8c0b625e792d2d2d6c193a73fbe6/jsoniter-scala-macros/shared/src/test/scala-3/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMakerNewKeywordSpec.scala#L62-L63) or [here](https://github.com/plokhotnyuk/jsoniter-scala/blob/8b730632c8ad8c0b625e792d2d2d6c193a73fbe6/jsoniter-scala-macros/shared/src/test/scala-3/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMakerNewKeywordSpec.scala#L62-L63). But I'm not sure that it can be reusable here. As example smithy4s just defines all codecs for basic types and derives for collections and complex ones like sum and product types: https://github.com/disneystreaming/smithy4s/blob/a5216184a5e584795c56a7a5a4def917849943f1/modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala

Reply:
Hmm it works in a simple isolated experiment:
```scala
//>using dep "com.github.plokhotnyuk.jsoniter-scala:jsoniter-scala-core_3:2.28.4"
//>using dep "com.github.plokhotnyuk.jsoniter-scala:jsoniter-scala-macros_3:2.28.4"
import com.github.plokhotnyuk.jsoniter_scala.macros._
import com.github.plokhotnyuk.jsoniter_scala.core._

case class MySimpletype2(v: String)
case class MySimpleType(a: Int, b: List[MySimpletype2])

object Main extends App {
  implicit val codec: JsonValueCodec[MySimpleType] = com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMaker.make[MySimpleType]
  val s = writeToString[MySimpleType](MySimpleType(1, List(MySimpletype2("a"))))
  println(s)
}
```

maybe it has something to do with location of the classes or the main CC codec :thinking:

Reply:
Or maybe it works only for Scala 3 :thinking:

Reply:
That example _should_ work, but I think if you tried `val s = writeToString[List[MySimpleType]](List(MySimpleType(1, List(MySimpletype2("a")))))` it would break. One thing I have only just appreciated is that jsoniter-scala doesn't need the intermediary serdes, so we needn't generate them for every schema model, only the top-level ones. Anyway, that's a different issue than the serdes defined here for primitive types, lists and options.

Reply:
Ah I understand now. Thanks for clarifying the difference @hughsimpson. The approach from Smithy4s looks interesting, thanks a lot for chiming in @plokhotnyuk. 
I think I am OK with current solution, looks like a good start and it could be optimized in the future.